server/src/graph_maker/tram_track_graph_transformer.py

```diff
@@ -0,0 +1,386 @@
+                node_a, dist_a = existing_points[i]
+                node_b, dist_b = existing_points[i + 1]
+
+                if dist_a > dist_b:
+                    node_a, dist_a, node_b, dist_b = node_b, dist_b, node_a, dist_a
+
+                segment_len = dist_b - dist_a
+                n = math.ceil(segment_len / max_distance)
+
+                if n == 1:
+                    continue
+
+                step = segment_len / n
+                prev_node = node_a
+                interpolated_nodes = []
+
+                for k in range(1, n):
+                    dist_current = dist_a + k * step
+                    if dist_current >= dist_b:
+                        _graph.add_edge(prev_node, node_b)
+                        break
+
+                    ip_3857 = line_3857.interpolate(dist_current)
+                    ip_wgs = transform(transformer_to_wgs, ip_3857)
+                    lat_new, lon_new = ip_wgs.y, ip_wgs.x
+
+                    existing = find_existing_node_by_coords(lat_new, lon_new)
+                    if existing:
+                        new_node_id = existing
+                    else:
+                        new_node_id = next_node_id
+                        next_node_id += 1
+                        _graph.add_node(
+                            new_node_id, lat=lat_new, lon=lon_new, ways=[way_id]
+                        )
+                    interpolated_nodes.append(new_node_id)
+
+                all_nodes = [node_a] + interpolated_nodes + [node_b]
+
+                if _graph.has_edge(node_a, node_b):
+                    edges_to_remove.add((node_a, node_b))
+                    for u, v in zip(all_nodes, all_nodes[1:]):
+                        _graph.add_edge(u, v, ways=[way_id])
+
+                if _graph.has_edge(node_b, node_a):
+                    edges_to_remove.add((node_b, node_a))
+                    for u, v in zip(reversed(all_nodes), list(reversed(all_nodes))[1:]):
+                        _graph.add_edge(u, v, ways=[way_id])
+
+        _graph.remove_edges_from(edges_to_remove)
+
+    def _initialize_graph(self):
+        self._graph = self._build_initial_graph()
+        self._permanent_nodes = self._find_permanent_nodes()
+        self._remove_unnecessary_nodes()
+        self._add_two_way_edges()
+        self._add_coordinates_from_data()
+        self._tag_nodes_with_way_ids()
+        self._build_dict()
+        while self._is_there_any_ways_to_merge():
+            self._merge_ways(tolerance=2.0)
+
+    def densify_graph_by_max_distance(
+        self, max_distance_in_meters: float
+    ) -> nx.DiGraph:
+        if max_distance_in_meters <= 0:
+            raise ValueError("max_distance_in_meters must be a positive float value")
+
+        graph_copy = self._graph.copy()
+        self._subdivide_ways(graph_copy, max_distance=max_distance_in_meters)
+        return graph_copy
+
+    @property
+    def permanent_node_ids(self):
+        return list(self._permanent_nodes)
```

Comment:
I think this way is more explicit as to what is going on here, also there's no shame in returning a set instead of a list.
```suggestion
        return self._permanent_nodes.copy()
```

Reply:
Good point 👍 - applied